src/browser/cdp/selectivity/utils.ts

```diff
@@ -0,0 +1,210 @@
+import { sortedIndex, memoize } from "lodash";
+import { SourceMapConsumer, type BasicSourceMapConsumer, type RawSourceMap } from "source-map";
+import fs from "fs";
+import path from "path";
+import { URL } from "url";
+import type { CDPRuntime } from "../domains/runtime";
+import type { CDPSessionId } from "../types";
+import { softFileURLToPath } from "../../../utils/fs";
+import type { NormalizedDependencies } from "./types";
+import { WEBPACK_PROTOCOL } from "./constants";
+
+export const fetchTextWithBrowserFallback = async (
+    url: string,
+    runtime: CDPRuntime,
+    sessionId: CDPSessionId,
+): Promise<string | Error> => {
+    const isSourceMapEmbedded = new URL(url).protocol === "data:";
+
+    if (isSourceMapEmbedded) {
+        // With "data" protocol it just decodes embedded source maps without actual network requests
+        // So we can do it directly from node.js
+        return fetch(url)
+            .then(r => r.text())
+            .catch((err: Error) => err);
+    }
+
+    // At first, trying to fetch sourceMaps directly from the node.js
+    // Then falling back to do it via browser (if Testplane has no direct network access to it, for example)
+    try {
+        return await fetch(url).then(r => r.text());
+    } catch {
+        return runtime
+            .evaluate<string>(sessionId, {
+                expression: `fetch("${url.replaceAll('"', '\\"')}").then(r => r.text())`,
+                awaitPromise: true,
+                returnByValue: true,
+            })
+            .then(r => r.result.value)
+            .catch((err: Error) => err);
+    }
+};
+
+const isWebpackProtocol = (posixRelativeSourceFilePath: string): boolean => {
+    return posixRelativeSourceFilePath.startsWith(WEBPACK_PROTOCOL);
+};
+
+export const patchSourceMapSources = (sourceMap: RawSourceMap, sourceRoot?: string): RawSourceMap => {
+    sourceMap.sourceRoot = sourceRoot || sourceMap.sourceRoot;
+
+    for (let i = 0; i < sourceMap.sources.length; i++) {
+        if (isWebpackProtocol(sourceMap.sources[i])) {
+            sourceMap.sources[i] = sourceMap.sources[i].slice(WEBPACK_PROTOCOL.length + 1);
+        }
+    }
+
+    return sourceMap;
+};
+
+export const extractSourceFilesDeps = async (
+    source: string,
+    sourceMaps: string,
+    startOffsets: number[],
+    sourceRoot: string,
+): Promise<Set<string>> => {
+    const dependantSourceFiles = new Set<string>();
+    const sourceMapsParsed = patchSourceMapSources(JSON.parse(sourceMaps), sourceRoot);
+
+    const consumer = (await new SourceMapConsumer(sourceMapsParsed)) as BasicSourceMapConsumer;
+
+    let sourceOffset = source.indexOf("\n");
+    const offsetToLine = [0];
+
+    while (sourceOffset !== -1) {
+        offsetToLine.push(++sourceOffset);
+        sourceOffset = source.indexOf("\n", sourceOffset);
+    }
```

Comment:
v8 coverage has straight "offsetToLine", while "source-maps" needs "line" and "column"
Here we are creating array, where "line" is index and "sourceOffset" is offset for the line